compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+			// Ensure algorithm is uppercased
+			algorithm = strings.ToUpper(algorithm)
+			tableCompressions[table][column] = algorithm
+
+			if _, ok := c.supportedAlgorithms[algorithm]; !ok {
+				return &UnsupportedCompressionError{
+					table:     table,
+					column:    column,
+					algorithm: algorithm,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The CompressionVerifier method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used and verified
+// for comparison equality
+func (c *CompressionVerifier) GetHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	return c.decompressAndHash(table, columns, rows, tableCompression)
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
+	quotedPK := quoteField(pkColumn)
+	sql, args, err := rowSelector(columns, pkColumn).
+		From(QuotedTableNameFromString(schema, table)).
+		Where(sq.Eq{quotedPK: pks}).
+		OrderBy(quotedPK).
+		ToSql()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// This query must be a prepared query. If it is not, querying will use
+	// MySQL's plain text interface, which will scan all values into []uint8
+	// if we give it []interface{}.
+	stmt, err := db.Prepare(sql)
+	if err != nil {
+		return nil, err
+	}
+
+	defer stmt.Close()
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+func (c *CompressionVerifier) decompressAndHash(table string, columns []schema.TableColumn, rows *sql.Rows, tableCompression map[string]string) (map[uint64][]byte, error) {
+	defer rows.Close()
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanGenericRow(rows, 2)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := rowData.GetUint64(0)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decodedRowData contains a map of all
+		// columns and associated decoded values by the index of the column
+		var decodedRowData map[int][]byte
+		for idx, column := range columns {
+			// Check if column is configured as compressed and decompress if necessary
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				decodedColData, err := c.Decode(table, column.Name, algorithm, rowData[idx].([]byte))
```

Comment:
nitpick: maybe `decodedCellData`, since this is only one cell (not the entire column)?

Reply:
Does calling it `Col`/`Column` make you think we're decompressing the entire column of the table and not the column bound by the row's PK/idx? I haven't seen any other libs refer to this as a `cell`, but if it's confusing we can certainly change it.